mlflow/tracking/client.py

```
@@ -1527,6 +1527,103 @@ def log_table(
+                extra_columns=["run_id", "parent_run_id"],
+            )
+        """
+        import pandas as pd
+
+        subset_tag_value = json.dumps({"path": artifact_file, "type": "table"})
+
+        # Build the filter string
+        tag_query = f"tags.{MLFLOW_LOGGED_ARTIFACTS} LIKE '%{subset_tag_value}%'"
+        run_query = ""
+        if run_ids is not None:
+            list_run_ids = ",".join(run_ids)
+            run_query = f" and attributes.run_id IN ('{list_run_ids}')"
+
+        filter_string = tag_query + run_query
+        runs = mlflow.search_runs(experiment_ids=[experiment_id], filter_string=filter_string)
+        data = pd.DataFrame()
+
+        # TODO: Add parallelism support here
+        def get_artifact_data(run):
+            run_id = run.run_id
+            norm_path = posixpath.normpath(artifact_file)
+            artifact_dir = posixpath.dirname(norm_path)
+            artifact_dir = None if artifact_dir == "" else artifact_dir
+            existing_predictions = pd.DataFrame()
+
+            artifacts = [f.path for f in self.list_artifacts(run_id, path=artifact_dir)]
+            if artifact_file in artifacts:
+                with tempfile.TemporaryDirectory() as tmpdir:
+                    downloaded_artifact_path = mlflow.artifacts.download_artifacts(
+                        run_id=run_id, artifact_path=artifact_file, dst_path=tmpdir
+                    )
+                    existing_predictions = pd.read_json(downloaded_artifact_path, orient="split")
+                    if extra_columns is not None:
+                        for column in extra_columns:
+                            existing_predictions[column] = existing_predictions.apply(
+                                lambda x, column=column: run[column], axis=1
+                            )
+            else:
+                raise MlflowException(f"Artifact {artifact_file} not found for run {run_id}.")
```

Can we validate all the tables are downloadable beforehand to fail fast and avoid useless downloads?

---

How do we verify that if the a particular artifact is downloadable? I thought, checking if the path exists should help with making sure that the file is downloadable.

Also this else should never ever happen if the search query is accurate. I have just added it for safety sake.

---

> Also this else should never ever happen if the search query is accurate. I have just added it for safety sake.

I meant something like this:

```python
for run_id in run_ids:
    artifacts = [
        f.path for f in self.list_artifacts(run_id, path=artifact_dir) if not f.is_dir
    ]
    raise MlflowException(f"Artifact {artifact_file} not found for run {run_id}.")


# Then, download them
```

> Also this else should never ever happen if the search query is accurate. I have just added it for safety sake.


I missed this.